### PR TITLE
add a octree2pointcloud conversion tool (does NOT depend on libpcl)

### PR DESCRIPTION
**Pull request by fferri**

### CODE REVIEW

File: octomap/src/octree2pointcloud.cpp [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  OcTree* tree = new OcTree(0.1);
+  if (!tree->readBinary(inputFilename)){
+    OCTOMAP_ERROR("Could not open file, exiting.\n");
+    exit(1);
+  }
+  
+  res = tree->getResolution();
+  std::vector<point3d> pcl;
+
+  for (OcTree::iterator it = tree->begin(res); it != tree->end(); ++it)
```
Comment (ahornung) [inline octomap/src/octree2pointcloud.cpp]: Argument for the iterator is the max_depth for traversal, not tree resolution. I would suggest to either get rid of it completely or allow the user to specify a max_depth from the command line.

You can then get completely rid of the `res` variable.

File: octomap/src/octree2pointcloud.cpp [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+void printUsage(char* self){
+  std::cerr << "USAGE: " << self << " <InputFile.bt> <OutputFile.pcd>\n";
+  std::cerr << "This tool creates a point cloud of the occupied cells\n";
+  exit(0);
+}
+
+
+int main(int argc, char** argv) {
+  // default values:
+  double res = 0.1;
```
Comment (ahornung) [inline octomap/src/octree2pointcloud.cpp]: See belo, should not be needed.

File: octomap/src/octree2pointcloud.cpp (hunk 1 above)
Comment (fferri) [inline octomap/src/octree2pointcloud.cpp]: My intention here was to traverse all the occupied voxel at the maximum resolution.

File: octomap/src/octree2pointcloud.cpp (hunk 2 above)
Comment (fferri) [inline octomap/src/octree2pointcloud.cpp]: How about   res = tree->getTreeDepth() ?

Comment (fferri) [inline octomap/src/octree2pointcloud.cpp]: of course  unsigned int res = tree->getTreeDepth();

File: octomap/src/octree2pointcloud.cpp (hunk 1 above)
Comment (ahornung) [inline octomap/src/octree2pointcloud.cpp]: You actually don't need res at all, the default behavior is to traverse all voxels at maximum resolution:
`it = tree->begin();`

Comment (fferri) [inline octomap/src/octree2pointcloud.cpp]: Ok, I'll change that.
But what if there are 8 occupied voxels, such that the parent voxel would be considered completely occupied?
Would the program return the center of the voxel at the upper level in that case?

**Comment by ahornung** [comment]
@fferri Ah, now I get why you wanted to specify the depth explicitly!

It is as you just wrote. When there is a pruned leaf node (compressed from 8 identical children) then only its center coordinate will be included in the point cloud, so it could look like holes. However, giving the leaf iterator the max tree depth won't change that, it only limits the maximum depth.

To avoid this behavior there are some options:
 * expand the complete tree (uncompress) to leafs at lowest level; easiest, but will consume a lot of memory
 * locally expand only the current leaf node when it's not at the lowest level, then compress again
 * Check the size of the current node and then create the 8/32/... points manually in the point cloud (compare with 2D pixel map projection in the octomap_server code). Won't consume more memory (except the target points), but more complicated to implement.

**Comment by fferri** [comment]
Could you please review fferri/octomap@e7e3d6e8a1792055f3dfd8836c677a3bd85d19a2 ?
I choose create a list of leaf occupied nodes which are not at maximum depth, and expand those. (not sure if the while is needed, but can't hurt).

It compiles, but I haven't tested it, because I need some scripts on my computer at work. (I will test it on Monday)

**Comment by ahornung** [comment]
Looks like a good solution, since you're only interested in the occupied nodes it makes sense to expand only those.

There's some room for improvement also: instead of the while loop you could collect the initial collapsed nodes in one sweep, and then expand them recursively until they reach the smallest resolution in a second sweep. In the first sweep, you could add the leafs at lowest level to the pcl vector, while in the second you would add the newly-expanded leafs to it. Might make a difference for large trees, but I don't think it's critical.

In terms of style, however, I suggest to replace the do/while with a regular while to conform with the other code in OctoMap.

**Pull request merged**